soteria-rust/lib/interp.ml

```diff
@@ -421,21 +421,34 @@ module Make (Heap : Heap_intf.S) = struct
         match (v1, v2) with
         | Base v1, Base v2 -> (
             match op with
-            | Ge | Gt | Lt | Le ->
-                let op =
-                  match op with
-                  | Ge -> Typed.geq
-                  | Gt -> Typed.gt
-                  | Lt -> Typed.lt
-                  | Le -> Typed.leq
-                  | _ -> assert false
-                in
+            | Ge | Gt | Lt | Le -> (
                 let* v1, v2, ty = cast_checked2 v1 v2 in
-                if Typed.equal_ty ty Typed.t_ptr then
-                  Heap.error `UBPointerComparison state
-                else
-                  let v = op v1 v2 |> Typed.int_of_bool in
-                  Result.ok (Base v, state)
+                match Typed.untype_type ty with
+                | Svalue.TInt ->
+                    let op =
+                      match op with
+                      | Ge -> Typed.geq
+                      | Gt -> Typed.gt
+                      | Lt -> Typed.lt
+                      | Le -> Typed.leq
+                      | _ -> assert false
```

Comment:
In general, I like to avoid assert false and use failwith "...." so that if something goes wrong because it is not actually unreachable, I have something more precise than assert false.